docs: add datatype vecf16 and svector (#54)
tensorchord/pgvecto.rs-docs: src/usage/vector-types.md

- `<=>` (`bvector_cos_ops`): cosine distance, defined as $1 - \frac{\Sigma x_iy_i}{\sqrt{\Sigma x_i^2 \Sigma y_i^2}}$.
- `<~>` (`bvector_jaccard_ops`): Jaccard distance, defined as $1 - \frac{|X\cap Y|}{|X\cup Y|}$.

Index can be created on `bvector` type as well.

```sql
CREATE INDEX your_index_name ON items USING vectors (embedding bvector_l2_ops);

SELECT * FROM items ORDER BY embedding <-> '[1,0,1]' LIMIT 5;
```

### Data type cast

Cast between `vector`:
```sql
SELECT '[1, 0, 1]'::vector::bvector;
SELECT '[1, 0, 1]'::bvector::vector;
```

From `ARRAY` or `real[]` to bvector:
```sql
SELECT ARRAY[1, 0, 1]::real[]::vector::bvector;
```

From string constructor:
```sql
SELECT '[1, 0, 1]'::bvector;
```

From binarize constructor:
```sql
SELECT binarize(ARRAY[-2, -1, 0, 1, 2]::real[]::vector);;
-- [0, 0, 0, 1, 1]

> ```

Reviewer: Not all pairs of vector types are cast-able.

We guarantee:
* casts between `vector` and `real[]` exist
* casts between `vector` and `[prefix]vector` exist
* casts between `[prefix]vector` and `[prefix]vec[suffix]` exist

And we provide more functions to build a vector:
* `to_[prefix]vec[suffix]`
* `binarize`

Every user-defined postgresql type can be derived from `text` so we do not have to emphasize it.

Index can be created on `bvector` type as well.

```sql
CREATE INDEX your_index_name ON items USING vectors (embedding bvector_l2_ops);

SELECT * FROM items ORDER BY embedding <-> '[1,0,1]' LIMIT 5;
```

### Data constructor

There is a binarize constructor to build `bvector` from `vector`:
```sql
SELECT binarize(ARRAY[-2, -1, 0, 1, 2]::real[]::vector);;
-- [0, 0, 0, 1, 1]
```

> 

Reviewer: Github makes style a mess.

- Word co-occurrence matrices
- Term frequency-inverse document frequency (TF-IDF) vectors
- User-item interaction matrices
- Network adjacency matrices

Sparse vectors in `pgvecto.rs` are called `svector`.

::: tip
`svector` is 32-bit float, 16-bit float sparse vector is not supported yet.
:::

Here's an example of creating a table with a svector column and inserting values:

```sql {3}
CREATE TABLE items (
  id bigserial PRIMARY KEY,
  embedding svector(10) NOT NULL
);

INSERT INTO items (embedding) VALUES ('[0.1,0,0,0,0,0,0,0,0,0]'), ('[0,0,0,0,0,0,0,0,0,0.5]');
```

We support three operators to calculate the distance between two `svector` values.

- `<->` (`svector_l2_ops`): squared Euclidean distance, defined as $\Sigma (x_i - y_i) ^ 2$.
- `<#>` (`svector_dot_ops`): negative dot product, defined as $- \Sigma x_iy_i$.
- `<=>` (`svector_cos_ops`): cosine distance, defined as $1 - \frac{\Sigma x_iy_i}{\sqrt{\Sigma x_i^2 \Sigma y_i^2}}$.

Index can be created on `svector` type as well.

```sql
CREATE INDEX your_index_name ON items USING vectors (embedding svector_l2_ops);

SELECT * FROM items ORDER BY embedding <-> '[0.3,0,0,0,0,0,0,0,0,0]' LIMIT 1;
```

> ### Data constructor
> 
> There is an index and value constructor to build arbitrary `svector`:
> ```sql
> SELECT to_svector(5, '{0,4}', '{0.3,0.5}');
> -- [0.3, 0, 0, 0, 0.5]
> ```

Reviewer: ```suggestion
There is also a function `to_svector` to build `svector`:
```sql
SELECT to_svector(5, '{0,4}', '{0.3,0.5}');
-- [0.3, 0, 0, 0, 0.5]
```

```

## `vecf16` half-precision vector

Stored as a half-precision number format, `vecf16` takes advantage of the 16-bit float, which uses half the memory and bandwidth compared to `vector`.
It is often faster than the regular `vector` datatype, but may lose some precision.

Here's an example of creating a table with a vecf16 column and inserting values:

```sql {3}
CREATE TABLE items (
  id bigserial PRIMARY KEY,
  embedding vecf16(3) NOT NULL
);

INSERT INTO items (embedding) VALUES ('[0.1, 0.2, 0]'), ('[0, 0.1, 0.2]');
```

We support three operators to calculate the distance between two `vecf16` values.

- `<->` (`vecf16_l2_ops`): squared Euclidean distance, defined as $\Sigma (x_i - y_i) ^ 2$.
- `<#>` (`vecf16_dot_ops`): negative dot product, defined as $- \Sigma x_iy_i$.
- `<=>` (`vecf16_cos_ops`): cosine distance, defined as $1 - \frac{\Sigma x_iy_i}{\sqrt{\Sigma x_i^2 \Sigma y_i^2}}$.

> Index can be created on `vecf16` type as well.

Reviewer: It's strange because we already list opclasses. We should put the statements before them.

```sql
CREATE INDEX your_index_name ON items USING vectors (embedding vecf16_l2_ops);

SELECT * FROM items ORDER BY embedding <-> '[0.3,0.2,0.1]' LIMIT 1;
```

> ## Vector type cast

Reviewer: ```suggestion
## Casts
```

For vector types, these casts are guaranteed:
- From `REAL[]` to `vector`
- From `vector` to other vector types

::: tip
Type casts between other vector types must be passed by `vector`.
:::

This diagram shows the conversion between different data types, where the types connected by arrows can be cast to each other:

> ```plantuml

Reviewer: [draw with quiver](https://q.uiver.app/#q=WzAsNyxbMywyLCJcXHRleHR7dmVjdG9yfSJdLFsyLDAsIlxcdGV4dHt2ZWNmMTZ9Il0sWzQsMCwiXFx0ZXh0e3ZlY2k4fSJdLFszLDUsIlxcdGV4dHtidmVjdG9yfSJdLFswLDIsIlxcdGV4dHtBUlJBWX0iXSxbNiwyLCJcXHRleHR7c3ZlY3Rvcn0iXSxbNywwLCJcXHRleHR7c3ZlY2YxNn0iXSxbMiwwXSxbMCwyXSxbMSwwXSxbMCwxXSxbMywwXSxbMCwzXSxbNCwwLCIiLDAseyJzdHlsZSI6eyJib2R5Ijp7Im5hbWUiOiJkb3R0ZWQifX19XSxbMCw1XSxbNSwwXSxbNiw1XSxbNSw2XSxbMCw0LCIiLDAseyJzdHlsZSI6eyJib2R5Ijp7Im5hbWUiOiJkb3R0ZWQifX19XV0=)

I prefer such a style.

@startuml
participant "REAL[]"

> participant ARRAY

Reviewer: Aren't they the same type?

![bvector](./images/bvector.png)

We can see that the `bvector`'s accuracy is not as good as the `vector` type, but it exceeds 95%  if we adopt adaptive retrieval.

## `svector` sparse vector

Different from dense vectors, sparse vectors are very high-dimensional but contain few non-zero values. Though you can treat them as traditional dense vectors, they can be calculated and stored much more efficiently by [some ways](https://en.wikipedia.org/wiki/Sparse_matrix).

Typically, sparse vectors could generated from:
- Word-word occurrence matrices
- Term frequency-inverse document frequency (TF-IDF) vectors
- User-item interaction matrices
- Network adjacency matrices

`pgvecto.rs` supports sparse vectors, it's called `svector`.

::: tip

> `svector` is 32-bit float, 16-bit float sparse vector is not supported now.

Reviewer: We can use ChatGPT to make the document clearer.

  embedding svector(10) NOT NULL
);

INSERT INTO items (embedding) VALUES ('[0.1,0,0,0,0,0,0,0,0,0]'), ('[0,0,0,0,0,0,0,0,0,0.5]');
```

We support three operators to calculate the distance between two `svector` values.

- `<->` (`svector_l2_ops`): squared Euclidean distance, defined as $\Sigma (x_i - y_i) ^ 2$.
- `<#>` (`svector_dot_ops`): negative dot product, defined as $- \Sigma x_iy_i$.
- `<=>` (`svector_cos_ops`): cosine distance, defined as $1 - \frac{\Sigma x_iy_i}{\sqrt{\Sigma x_i^2 \Sigma y_i^2}}$.

Index can be created on `svector` type as well.

```sql
CREATE INDEX your_index_name ON items USING vectors (embedding svector_l2_ops);

SELECT * FROM items ORDER BY embedding <-> '[0.3,0,0,0,0,0,0,0,0,0]' LIMIT 1;
```

### Data type cast

Cast between `vector`:
```sql
SELECT '[0.3, 0, 0, 0, 0.5]'::vector::svector;
SELECT '[0.3, 0, 0, 0, 0.5]'::svector::vector;
```

From `ARRAY` or `real[]` to svector:
```sql
SELECT ARRAY[random(), 0, 0, 0, 0.5]::real[]::vector::svector;
```

From string constructor:
```sql
SELECT '[0.3, 0, 0, 0, 0.5]'::svector;
```

From index and value constructor:
```sql
SELECT to_svector(5, '{0,4}', '{0.3,0.5}');
-- [0.3, 0, 0, 0, 0.5]

> ```

Reviewer: We do not have to list these for every vector-type.

Here's an example of creating a table with a vecf16 column and inserting values:

```sql {3}
CREATE TABLE items (
  id bigserial PRIMARY KEY,
  embedding vecf16(3) NOT NULL
);

INSERT INTO items (embedding) VALUES ('[0.1, 0.2, 0]'), ('[0, 0.1, 0.2]');
```

We support three operators to calculate the distance between two `vecf16` values.

- `<->` (`vecf16_l2_ops`): squared Euclidean distance, defined as $\Sigma (x_i - y_i) ^ 2$.
- `<#>` (`vecf16_dot_ops`): negative dot product, defined as $- \Sigma x_iy_i$.
- `<=>` (`vecf16_cos_ops`): cosine distance, defined as $1 - \frac{\Sigma x_iy_i}{\sqrt{\Sigma x_i^2 \Sigma y_i^2}}$.

Index can be created on `vecf16` type as well.

```sql
CREATE INDEX your_index_name ON items USING vectors (embedding vecf16_l2_ops);

SELECT * FROM items ORDER BY embedding <-> '[0.3,0.2,0.1]' LIMIT 1;
```

### Data type cast

Cast between `vector`:
```sql
SELECT '[0.3, 0.2, 0.1]'::vector::vecf16;
SELECT '[0.3, 0.2, 0.1]'::vecf16::vector;
```

From `ARRAY` or `real[]` to vecf16:
```sql
SELECT ARRAY[random(), 0, 0.1]::real[]::vector::vecf16;
```

From string constructor:
```sql
SELECT '[0.3, 0.2, 0.1]'::vecf16;

> ```

Reviewer: Same.
